src/string_id.h

```diff
@@ -209,29 +312,40 @@ class string_id
         }
 
     private:
-        std::string _id;
-        // cached int_id counterpart of this string_id
-        mutable int _cid;
         // generic_factory version that corresponds to the _cid
-        mutable int64_t _version;
+        mutable int64_t _version = INVALID_VERSION;
+        // cached int_id counterpart of this string_id
+        mutable int _cid = INVALID_CID;
+        // structure that captures the actual "identity" of this string_id
+        Identity _id;
 
         inline void set_cid_version( int cid, int64_t version ) const {
             _cid = cid;
             _version = version;
         }
 
         friend class generic_factory<T>;
+        friend struct std::hash<string_id<T>>;
 };
 
 // Support hashing of string based ids by forwarding the hash of the string.
 namespace std
 {
 template<typename T>
-struct hash< string_id<T> > {
+struct hash<string_id<T>> {
     std::size_t operator()( const string_id<T> &v ) const noexcept {
-        return hash<std::string>()( v.str() );
+        using IdType = decltype( v._id._id );
+        return std::hash<IdType>()( v._id._id );
     }
 };
 } // namespace std
 
+/** Lexicographic order comparator for string_ids */
+template<typename T>
+struct lexicographic : std::binary_function<string_id<T>, string_id<T>, bool> {
```

Comment:
`std::binary_function` is deprecated; you shouldn't need to use it here.

(Also, as mentioned above, I think we should be able to delete this entirely; I doubt we're ever intentionally sorting `string_id`s lexicographically)